lib/kube_link_generator.rb

```diff
@@ -0,0 +1,116 @@
+require 'kubeclient'
+require 'uri'
+require_relative 'exceptions'
+
+# KubeLinkSpecs provides the information required to generate BOSH links by
+# pretending to be a hash.
+class KubeLinkSpecs
+  # ANNOTATION_AZ is the Kube annotation for the (availability) zone
+  ANNOTATION_AZ = 'failure-domain.beta.kubernetes.io/zone'.freeze
+
+  def initialize(spec, namespace, kube_client)
+    @links = {}
+    @client = kube_client
+    @namespace = namespace
+    @spec = spec || {}
+  end
+
+  def this_name
+    @spec['job']['name']
+  end
+
+  def pod_index(name)
+    index = name.rpartition('-').last
+    return index.to_i if /^\d+$/ =~ index
+    chars = 'bcdfghjklmnpqrstvwxz0123456789'
+    index.chars.map { |c| chars.index(c) }.reduce(0) { |v, c| v * chars.length + c }
+  end
+
+  def get_pods_for_role(role_name, wait_for_ip)
+    loop do
+      1.times do
+        pods = @client.get_pods(namespace: @namespace, label_selector: "skiff-role-name=#{role_name}")
+        if wait_for_ip
+          # Wait until all pods have IP addresses and properties
+          break unless pods.all? { |pod| pod.status.podIP }
+          break unless pods.all? { |pod| pod.metadata.annotations['skiff-exported-properties'] }
+        else
+          # We just need one pod with exported properties
+          pods.select! { |pod| pod.status.podIP }
+          pods.select! { |pod| pod.metadata.annotations['skiff-exported-properties'] }
+        end
+        return pods unless pods.empty?
+      end
+      sleep 1
+    end
+  end
+
+  def get_pod_instance_info(pod, job)
+    index = pod_index(pod.metadata.name)
+    properties = JSON.parse(pod.metadata.annotations['skiff-exported-properties'])
+    {
+      'name' => pod.metadata.name,
+      'index' => index,
+      'id' => pod.metadata.name,
+      'az' => pod.metadata.annotations['failure-domain.beta.kubernetes.io/zone'] || 'az0',
+      'address' => pod.status.podIP,
+      'properties' => properties.fetch(job, {}),
+      'bootstrap' => index.zero?
+    }
+  end
+
+  def get_svc_instance_info(role_name, job)
+    svc = @client.get_service(role_name, @namespace)
+    pod = get_pods_for_role(role_name, false).first
+    properties = JSON.parse(pod.metadata.annotations['skiff-exported-properties'])
+    {
+      'name' => svc.metadata.name,
+      'index' => 0,
```

Comment:
There are no indices for services (because they forward to actual things); the 0 here is just completely made up.

Comment added in dba6526399ca4d0976e352ea8f657d0fcacc0c1